include/utils.h

```diff
@@ -0,0 +1,196 @@
+             const RefinementParameters &parameters = RefinementParameters())
+{
+  Assert(
+    (dynamic_cast<parallel::TriangulationBase<reduced_dim, spacedim> *>(
+       &embedded_triangulation) == nullptr),
+    ExcMessage(
+      "The embedded triangulation must not be distributed. It will be partitioned later."));
+
+  namespace bgi = boost::geometry::index;
+
+  // build caches so that we can get local trees
+  GridTools::Cache<spacedim, spacedim>    space_cache{space_triangulation};
+  GridTools::Cache<reduced_dim, spacedim> embedded_cache{
+    embedded_triangulation};
+
+  auto refine = [&]() {
+    bool done        = false;
+    bool global_done = false;
+
+    double min_embedded = 1e10;
+    double max_embedded = 0;
+    double min_space    = 1e10;
+    double max_space    = 0;
+
+    while (global_done == false)
+      {
+        done = true;
+
+        // Bounding boxes of the space grid
+        const auto &tree =
+          space_cache.get_locally_owned_cell_bounding_boxes_rtree();
+
+        // Bounding boxes of the embedded grid
+        const auto &embedded_tree =
+          embedded_cache.get_cell_bounding_boxes_rtree();
+
+        // Let's check all cells whose bounding box contains an embedded
+        // bounding box
+        const bool use_space    = parameters.use_space;
+        const bool use_embedded = parameters.use_embedded;
+
+        AssertThrow(!(use_embedded && use_space),
+                    ExcMessage("You can't refine both the embedded and "
+                               "the space grid at the same time."));
+
+        for (const auto &[embedded_box, embedded_cell] : embedded_tree)
+          {
+            const auto &[p1, p2] = embedded_box.get_boundary_points();
+            const auto diameter  = p1.distance(p2);
+            min_embedded         = std::min(min_embedded, diameter);
+            max_embedded         = std::max(max_embedded, diameter);
+
+            for (const auto &[space_box, space_cell] :
+                 tree | bgi::adaptors::queried(bgi::intersects(embedded_box)))
+              {
+                const auto &[sp1, sp2]    = space_box.get_boundary_points();
+                const auto space_diameter = sp1.distance(sp2);
+                min_space                 = std::min(min_space, space_diameter);
+                max_space                 = std::max(max_space, space_diameter);
+
+                if (use_embedded && space_diameter < diameter)
+                  {
+                    embedded_cell->set_refine_flag();
+                    done = false;
+                  }
+                if (use_space && diameter < space_diameter)
+                  {
+                    space_cell->set_refine_flag();
+                    done = false;
+                  }
+              }
+          }
+
+        // Synchronize done variable across all processes, otherwise we might
+        // deadlock
+        global_done =
+          Utilities::MPI::min(static_cast<int>(done),
```

Comment:
Compared to the serial case, we need this to add this to avoid possible deadlocks.